File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/gui/providers/GUIController.java
package com.gmail.borlandlp.minigamesdtools.arena.gui.providers;

import com.gmail.borlandlp.minigamesdtools.arena.ArenaComponent;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPhaseComponent;

import java.util.ArrayList;
import java.util.List;

public class GUIController extends ArenaComponent implements ArenaPhaseComponent {
    private List<GUIProvider> providers = new ArrayList<>();
    private GUIListener guiListener;

    public List<GUIProvider> getProviders() {
        return providers;
    }

    public void addProvider(GUIProvider provider) {
        this.providers.add(provider);
    }

    public void removeProvider(GUIProvider provider) {
        this.providers.remove(provider);
    }

    public void update() {

    }

    @Override
    public void beforeRoundStarting() {

    }

    @Override
    public void onRoundEnd() {

    }

    @Override
    public void onInit() {
        this.guiListener = new GUIListener(this);
        try {
            this.getArena().getEventAnnouncer().register(this.guiListener);
        } catch (Exception e) {
            e.printStackTrace();
        }

        for (ArenaPhaseComponent provider : this.getProviders()) {
            try {
                this.getArena().getPhaseComponentController().register(provider);
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    public void beforeGameStarting() {

    }

    public void gameEnded() {
        this.getArena().getEventAnnouncer().unregister(this.guiListener);
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/gui/providers/GUIController.kt
package com.gmail.borlandlp.minigamesdtools.arena.gui.providers

import com.gmail.borlandlp.minigamesdtools.arena.ArenaComponent
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPhaseComponent
import java.util.*

class GUIController : ArenaComponent(), ArenaPhaseComponent {
    private val providers: MutableList<GUIProvider> = ArrayList()
    private var guiListener: GUIListener? = null

    fun getProviders(): List<GUIProvider> {
        return providers
    }

    fun addProvider(provider: GUIProvider) {
        providers.add(provider)
    }

    fun removeProvider(provider: GUIProvider?) {
        providers.remove(provider)
    }

    override fun update() {}
    override fun beforeRoundStarting() {}
    override fun onRoundEnd() {}
    override fun onInit() {
        guiListener = GUIListener(this)
        try {
            arena!!.eventAnnouncer.register(guiListener!!)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        for (provider in getProviders()) {
            try {
                arena!!.phaseComponentController.register(provider)
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }
    }

    override fun beforeGameStarting() {}
    override fun gameEnded() {
        arena!!.eventAnnouncer.unregister(guiListener!!)
    }
}